platforms/software/testing-base-infrastructure/src/main/java/org/gradle/api/internal/tasks/testing/filter/TestSelectionMatcher.java

```diff
@@ -36,28 +38,69 @@
  * </ul>
  *
  * In both cases, if the pattern starts with an upper-case letter, it will be used to match
- * simple class name;
- * otherwise, it will be used to match full qualified class name.
+ * the simple class name;
+ * otherwise, it will be used to match the fully qualified class name.
  */
 public class TestSelectionMatcher {
-    private final List<TestPattern> buildScriptIncludePatterns;
-    private final List<TestPattern> buildScriptExcludePatterns;
-    private final List<TestPattern> commandLineIncludePatterns;
+    private final List<ClassTestPattern> buildScriptIncludePatterns;
+    private final List<ClassTestPattern> buildScriptExcludePatterns;
+    private final List<ClassTestPattern> commandLineIncludePatterns;
+
+    private final List<FileTestPattern> fileBuildScriptIncludePatterns;
+    private final List<FileTestPattern> fileBuildScriptExcludePatterns;
+    private final List<FileTestPattern> fileCommandLineIncludePatterns;
```

Comment:
I went with a simple duplication approach here.  There is probably a way to consolidate some of the code here, but I'm not too concerned with the 3 extra fields - it helps ensure I haven't changed the other behavior.